### ISSUE TITLE
scaler transformed data

### ISSUE THREAD

**Issue by kashif** [thuml/Nonstationary_Transformers#8]
Hello! 

If I am not mistaken, you preprocess the time series right at the start with std scaling and then apply this method and I was wondering, does that mean your calculated mu and std would then be close to 0 and 1 respectively? Would you know if this is the case? If you take context windows, due to the std scaling the mu and std would be close to 0 and 1 no?

Also, I believe the metrics due to the std scaling in your table and those of all the papers you compare with are in reality the NMSE and NMAE metrics rather than MSE and MAE

Finally, if one calculates the mu and std and provides these as static covariates to the transformer do you think there is a need for the projection networks?

I have a version of the Std-Scaling, if you are interested (together with static covariates support for mu and std), here https://github.com/kashif/pytorch-transformer-ts/blob/main/ns-transformer/ns-transformer-electricity.ipynb 

Looking forward to your insights!

**Comment by WenWeiTHU** [comment]
Hi, thanks for your interest.

* About **preprocess the time series right at the start with std scaling**: Do you mean the `StandardScaler` in `dataloader.py`? If so, the difference is that the `StandardScaler` makes **the whole series** follow mu=0 and std=1, while our normalization makes **the series from each sliding window** have mu=0 and std=1. Even the the whole series is normalized firstly, the calculated mu and std in each window (model’s raw input) will not close to 0 and 1. (See #9 for the visualization)

* About the **metrics**: of course! the metrics are `Normed MSE` and `Normed MAE` as you state, and these metrics and data processing are consistent with most time series forecasting benchmark of depth models at present, such as FEDformer, Pyraformer and Autoformer.

* About **provided mu and std as covariates**: I think pre-computed(estimated) statistics will be helpful to replace the projection networks for De-stationary Attention in Eq(5). The question is how to obtained it, could you please provide some details?
![image](https://user-images.githubusercontent.com/39073236/202097150-cef54590-e9e9-43c8-9b91-51db3ebe9136.png)

* A very nice repo! We will keep following it.

**Comment by kashif** [comment]
Regarding your last point and question, the way I do it is to take the mu and sigma from the StdScaler for each context window of the raw (without the StandardScaler) time series, and then concat them to the covariates vector as a static covariate which then passed to the standard Attention module:

https://github.com/kashif/pytorch-transformer-ts/blob/main/ns-transformer/module.py#L320

Also, note in your paper you mentioned DAIN, Rev-IN as related, however in DeepAR they also scale by the mean abs of the raw window and also pass this scaling factor to the RNN again as a static real-valued covariate `scale.log()`:

https://github.com/awslabs/gluonts/blob/dev/src/gluonts/torch/model/deepar/module.py#L240

In fact, I am doing this for all the models in my implementation as this is handled in the feature creation part which is the same for all models. I believe in this way the models are also aware of the distribution regime they are learning/forecasting over, but I do not think they wrote this in their paper explicitly as far as I remember. I hope this helps!
